File: api/src/models/icons.py
import re
from enum import Enum
from pydantic import BaseModel

ICON_WORD_BOUNDARY = re.compile(r"([a-z0-9])([A-Z])")
ICON_NON_SLUG = re.compile(r"[^a-zA-Z0-9]+")


class Icon(str, Enum):
    """Lucide icon slugs supported by the web runtime."""

    ACTIVITY = "activity"
    ARROW_RIGHT = "arrow-right"
    BANKNOTE = "banknote"
    BELL = "bell"
    BOX = "box"
    BOXES = "boxes"
    BUILDING_2 = "building-2"
    CHECK = "check"
    CLIPBOARD_LIST = "clipboard-list"
    CONTAINER = "container"
    CPU = "cpu"
    DATABASE = "database"
    DOWNLOAD = "download"
    HARD_DRIVE = "hard-drive"
    LAYERS = "layers"
    LAYOUT_DASHBOARD = "layout-dashboard"
    LAYOUT_GRID = "layout-grid"
    LINK = "link"
    LIST = "list"
    LIST_CHECK = "list-check"
    MAP_PIN = "map-pin"
    PLUS = "plus"
    ROCKET = "rocket"
    ROTATE_CCW = "rotate-ccw"
    SETTINGS_2 = "settings-2"
    SHIELD_CHECK = "shield-check"
    SLIDERS_HORIZONTAL = "sliders-horizontal"
    TIMER = "timer"
    USERS = "users"
    X = "x"


class IconCatalog(BaseModel):
    """List supported Lucide icon slugs."""

    # Catalog
    icons: list[Icon]


def parse_icon(icon: str | Icon | None) -> Icon | None:
    """Return a normalized icon enum member or reject unknown slugs."""

    if icon is None:
        return None

    if isinstance(icon, Icon):
        normalized_icon = icon.value
    else:
        normalized_icon = ICON_WORD_BOUNDARY.sub(r"\1-\2", icon.strip())
        normalized_icon = ICON_NON_SLUG.sub("-", normalized_icon).lower().strip("-")

    if not normalized_icon:
        return None

    try:
        return Icon(normalized_icon)
    except ValueError as exc:
        raise ValueError("Icon must be a supported Lucide icon slug") from exc

File: web/src/lib/icons.ts
import {
    Activity,
    ArrowRight,
    Banknote,
    Bell,
    Box,
    Boxes,
    Building2,
    Check,
    ClipboardList,
    Container,
    Cpu,
    Database,
    Download,
    HardDrive,
    Layers,
    LayoutDashboard,
    LayoutGrid,
    Link as LinkIcon,
    List as ListIcon,
    ListChecks,
    MapPin,
    Plus,
    Rocket,
    RotateCcw,
    Settings2,
    ShieldCheck,
    SlidersHorizontal,
    Timer,
    Users,
    X,
    type LucideIcon,
} from 'lucide-react';

export const ICON_NAMES = [
    'activity',
    'arrow-right',
    'banknote',
    'bell',
    'box',
    'boxes',
    'building-2',
    'check',
    'clipboard-list',
    'container',
    'cpu',
    'database',
    'download',
    'hard-drive',
    'layers',
    'layout-dashboard',
    'layout-grid',
    'link',
    'list',
    'list-check',
    'map-pin',
    'plus',
    'rocket',
    'rotate-ccw',
    'settings-2',
    'shield-check',
    'sliders-horizontal',
    'timer',
    'users',
    'x',
] as const;

export type IconName = (typeof ICON_NAMES)[number];

const ICON_NAME_SET = new Set<string>(ICON_NAMES);

const ICON_COMPONENTS: Record<IconName, LucideIcon> = {
    activity: Activity,
    'arrow-right': ArrowRight,
    banknote: Banknote,
    bell: Bell,
    box: Box,
    boxes: Boxes,
    'building-2': Building2,
    check: Check,
    'clipboard-list': ClipboardList,
    container: Container,
    cpu: Cpu,
    database: Database,
    download: Download,
    'hard-drive': HardDrive,
    layers: Layers,
    'layout-dashboard': LayoutDashboard,
    'layout-grid': LayoutGrid,
    link: LinkIcon,
    list: ListIcon,
    'list-check': ListChecks,
    'map-pin': MapPin,
    plus: Plus,
    rocket: Rocket,
    'rotate-ccw': RotateCcw,
    'settings-2': Settings2,
    'shield-check': ShieldCheck,
    'sliders-horizontal': SlidersHorizontal,
    timer: Timer,
    users: Users,
    x: X,
};

/** Returns whether a string is a supported icon slug. */
export function isIconName(name: string): name is IconName {
    return ICON_NAME_SET.has(name);
}

/** Resolves supported icon names to Lucide components. */
export function getIconComponent(name: string): LucideIcon | undefined {
    return isIconName(name) ? ICON_COMPONENTS[name] : undefined;
}
